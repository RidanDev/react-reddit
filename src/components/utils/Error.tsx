import React, { FunctionComponent } from 'react';
import { createStyles, withStyles, WithStyles, Theme } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import ReportIcon from '@material-ui/icons/Report';
import classNames from 'classnames';

const styles = (theme: Theme) =>
  createStyles({
    error: {
      display: 'flex',
      flexGrow: 1,
      alignItems: 'center',
      justifyContent: 'center',
      flexDirection: 'column',
    },
    gutters: {
      ...theme.mixins.gutters(),
      paddingTop: theme.spacing.unit * 2,
      paddingBottom: theme.spacing.unit * 2,
    },
    text: {
      marginTop: theme.spacing.unit,
      marginBottom: theme.spacing.unit,
    },
  });

export interface IErrorProps extends WithStyles<typeof styles> {
  message?: string;
  buttonMessage?: string;
  gutters?: boolean;
  onButtonClick?(): void;
}

const Error: FunctionComponent<IErrorProps> = React.memo(
  ({ classes, onButtonClick, message, buttonMessage, gutters = false }) => {
    return (
      <div
        className={classNames(classes.error, {
          [classes.gutters]: gutters,
        })}
      >
        <ReportIcon style={{ fontSize: 80 }} color="disabled" />
        <Typography variant="body1" className={classes.text}>
          {message || 'Oops, something went wrong'}
        </Typography>
        <Button color="primary" onClick={onButtonClick}>
          {buttonMessage || 'Retry'}
        </Button>
      </div>
    );
  },
);

export default withStyles(styles)(Error);
