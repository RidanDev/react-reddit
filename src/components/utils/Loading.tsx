import React, { FunctionComponent } from 'react';
import { createStyles, makeStyles } from '@material-ui/styles';
import CircularProgress from '@material-ui/core/CircularProgress';

const useStyles = makeStyles(
  createStyles({
    loading: {
      display: 'flex',
      flexGrow: 1,
      alignItems: 'center',
      justifyContent: 'center',
    },
  }),
);

const Loading: FunctionComponent<{}> = React.memo(() => {
  const classes = useStyles();

  return (
    <div className={classes.loading}>
      <CircularProgress />
    </div>
  );
});

export default Loading;
