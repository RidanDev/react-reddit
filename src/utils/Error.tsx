import React, { FunctionComponent } from 'react';
import { createStyles, withStyles, WithStyles, Theme } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import ReportIcon from '@material-ui/icons/Report';

const styles = (theme: Theme) =>
  createStyles({
    error: {
      display: 'flex',
      flexGrow: 1,
      alignItems: 'center',
      justifyContent: 'center',
      flexDirection: 'column',
    },
    text: {
      marginTop: theme.spacing.unit,
      marginBottom: theme.spacing.unit,
    },
  });

export interface IErrorProps extends WithStyles<typeof styles> {
  onRetryClicked?(): void;
}

const Error: FunctionComponent<IErrorProps> = ({ classes, onRetryClicked }) => {
  return (
    <div className={classes.error}>
      <ReportIcon style={{ fontSize: 80 }} color="disabled" />
      <Typography variant="body1" className={classes.text}>
        Oops, something went wrong
      </Typography>
      <Button color="primary" onClick={onRetryClicked}>
        Retry
      </Button>
    </div>
  );
};

export default withStyles(styles)(Error);
