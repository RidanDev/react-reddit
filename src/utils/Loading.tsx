import React, { FunctionComponent } from 'react';
import { createStyles, withStyles, WithStyles } from '@material-ui/core/styles';
import CircularProgress from '@material-ui/core/CircularProgress';

const styles = () =>
  createStyles({
    loading: {
      display: 'flex',
      flexGrow: 1,
      alignItems: 'center',
      justifyContent: 'center',
    },
  });

export interface ILoadingProps extends WithStyles<typeof styles> {}

const Loading: FunctionComponent<ILoadingProps> = React.memo(({ classes }) => {
  return (
    <div className={classes.loading}>
      <CircularProgress />
    </div>
  );
});

export default withStyles(styles)(Loading);
