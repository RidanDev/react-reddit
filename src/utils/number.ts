const mapping = [
  { number: 1000000000, postfix: 'b' },
  { number: 1000000, postfix: 'm' },
  { number: 1000, postfix: 'k' },
];

export interface IHumanizeOptions {
  /**
   * Adds a space between the number and the letter
   */
  space?: boolean;
  /**
   * Transform the letter to uppercase
   */
  upperCase?: boolean;
  /**
   * The number of decimal places that have to be preserved
   */
  precision?: number;
}

export function humanize(
  num: number,
  { space, upperCase, precision = 0 }: IHumanizeOptions = {},
): string {
  for (const map of mapping) {
    let fixedNumber: number | string = num / map.number;
    // We do this to prevent things like `1.0k`
    fixedNumber = Number.isInteger(fixedNumber) ? fixedNumber : fixedNumber.toFixed(precision);
    // If the number given is smaller than the current divider we have to continue to the next.
    // However, we wanna handle cases like `999` being transformed to `1k`
    if (map.number > num && Number(fixedNumber) < 1) continue;

    return `${fixedNumber}${space ? ' ' : ''}${
      upperCase ? map.postfix.toUpperCase() : map.postfix
    }`;
  }
  return num.toString();
}
