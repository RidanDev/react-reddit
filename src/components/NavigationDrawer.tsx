import React, { FunctionComponent } from 'react';
import { createStyles, withStyles, WithStyles } from '@material-ui/core/styles';
import withWidth, { WithWidth, isWidthDown } from '@material-ui/core/withWidth';
import Drawer from '@material-ui/core/Drawer';
import List from '@material-ui/core/List';
import { Location } from '@reach/router';
import Loading from './utils/Loading';
import Error from './utils/Error';
import NavigationDrawerItem from './NavigationDrawerItem';
import { drawerWidth } from '../themes';
import useReddit from '../hooks/useReddit';

const styles = () =>
  createStyles({
    drawer: {
      width: drawerWidth,
      flexShrink: 0,
    },
    drawerPaper: {
      width: drawerWidth,
      paddingTop: 48,
    },
  });

export interface INavigationDrawerProps extends WithStyles<typeof styles>, WithWidth {
  drawerOpen?: boolean;
  onDrawerClosed?: () => void;
}

const NavigationDrawer: FunctionComponent<INavigationDrawerProps> = ({
  classes,
  drawerOpen = true,
  onDrawerClosed,
  width,
}) => {
  const { loading, error, data, retry } = useReddit('/subreddits/default');

  return (
    <Location>
      {({ location }) => (
        <Drawer
          className={classes.drawer}
          variant={isWidthDown('sm', width) ? 'temporary' : 'persistent'}
          anchor="left"
          open={drawerOpen}
          onClose={onDrawerClosed}
          classes={{ paper: classes.drawerPaper }}
          ModalProps={{ keepMounted: true }}
          data-testid="navigation-drawer"
        >
          {loading && <Loading />}
          {error && <Error onButtonClick={retry} />}
          {data ? (
            <List dense={true} component="div">
              {data.map((subreddit: any) => (
                <NavigationDrawerItem
                  key={subreddit.id}
                  subreddit={subreddit}
                  isActive={location.pathname === subreddit.url}
                />
              ))}
            </List>
          ) : null}
        </Drawer>
      )}
    </Location>
  );
};

export default withWidth()(withStyles(styles)(NavigationDrawer));
