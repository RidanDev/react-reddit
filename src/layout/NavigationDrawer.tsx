import React, { FunctionComponent } from 'react';
import { createStyles, withStyles, WithStyles } from '@material-ui/core/styles';
import withWidth, { WithWidth, isWidthDown } from '@material-ui/core/withWidth';
import Drawer from '@material-ui/core/Drawer';
import Loading from '../utils/Loading';
import Error from '../utils/Error';
import { drawerWidth } from '../themes';
import useReddit from '../hooks/useReddit';

const styles = () =>
  createStyles({
    drawer: {
      width: drawerWidth,
      flexShrink: 0,
    },
    drawerPaper: {
      width: drawerWidth,
      paddingTop: 48,
    },
  });

export interface INavigationDrawerProps extends WithStyles<typeof styles>, WithWidth {
  drawerOpen?: boolean;
  onDrawerClosed?: () => void;
}

const NavigationDrawer: FunctionComponent<INavigationDrawerProps> = ({
  classes,
  drawerOpen = true,
  onDrawerClosed,
  width,
}) => {
  const { loading, error, retry } = useReddit('subreddits/default');

  return (
    <Drawer
      className={classes.drawer}
      variant={isWidthDown('sm', width) ? 'temporary' : 'persistent'}
      anchor="left"
      open={drawerOpen}
      onClose={onDrawerClosed}
      classes={{ paper: classes.drawerPaper }}
      ModalProps={{ keepMounted: true }}
      data-testid="navigation-drawer"
    >
      {loading && <Loading />}
      {error && <Error onRetryClicked={retry} />}
    </Drawer>
  );
};

export default withWidth()(withStyles(styles)(NavigationDrawer));
