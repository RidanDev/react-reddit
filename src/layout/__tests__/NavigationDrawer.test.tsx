import React from 'react';
import nock from 'nock';
import { renderWithTheme } from '../../../test/utils';
import NavigationDrawer from '../NavigationDrawer';

describe('<NavigationDrawer />', () => {
  const data = {
    data: {
      children: [{ name: '/r/all' }, { name: '/r/best' }, { name: '/r/top' }],
    },
  };

  beforeEach(() => {
    nock('https://api.reddit.com')
      .get('/subreddits/default')
      .reply(200, data);
  });

  it.each(['md', 'lg', 'xl'])('should be visible even when closed with %s width', width => {
    const { getByTestId } = renderWithTheme(
      <NavigationDrawer width={width} drawerOpen={false} />,
      'light',
    );
    expect(getByTestId('navigation-drawer')).toBeVisible();
  });

  it.each(['sm', 'xs'])('should not be visible when closed with %s width', width => {
    const { getByTestId } = renderWithTheme(
      <NavigationDrawer width={width} drawerOpen={false} />,
      'light',
    );
    expect(getByTestId('navigation-drawer')).not.toBeVisible();
  });

  it.each(['xs', 'sm', 'md', 'lg', 'xl'])('should be visible when opened with %s width', size => {
    const { getByTestId } = renderWithTheme(<NavigationDrawer width={size} drawerOpen />, 'light');
    expect(getByTestId('navigation-drawer')).toBeVisible();
  });

  it('should switch themes', () => {
    const { container, rerenderWithTheme } = renderWithTheme(
      <NavigationDrawer width="xl" drawerOpen />,
      'light',
    );
    expect(container.firstChild).toMatchSnapshot();
    rerenderWithTheme(<NavigationDrawer width="xl" drawerOpen />, 'dark');
    expect(container.firstChild).toMatchSnapshot();
  });
});
