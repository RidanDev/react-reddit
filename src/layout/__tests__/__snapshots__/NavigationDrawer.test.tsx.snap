// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavigationDrawer /> should switch themes 1`] = `
<div
  class="MuiDrawer-root-539 MuiDrawer-docked-540 NavigationDrawer-drawer-537"
  data-testid="navigation-drawer"
>
  <div
    class="MuiPaper-root-551 MuiPaper-elevation0-553 MuiDrawer-paper-541 NavigationDrawer-drawerPaper-538 MuiDrawer-paperAnchorLeft-542 MuiDrawer-paperAnchorDockedLeft-546"
    direction="right"
  >
    <div
      class="Component-loading-578"
    >
      <div
        class="MuiCircularProgress-root-579 MuiCircularProgress-colorPrimary-582 MuiCircularProgress-indeterminate-581"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg-584"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle-585 MuiCircularProgress-circleIndeterminate-587"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`<NavigationDrawer /> should switch themes 2`] = `
<div
  class="MuiDrawer-root-616 MuiDrawer-docked-617 NavigationDrawer-drawer-628"
  data-testid="navigation-drawer"
>
  <div
    class="MuiPaper-root-589 MuiPaper-elevation0-591 MuiDrawer-paper-618 NavigationDrawer-drawerPaper-629 MuiDrawer-paperAnchorLeft-619 MuiDrawer-paperAnchorDockedLeft-623"
    direction="right"
  >
    <div
      class="Component-loading-640"
    >
      <div
        class="MuiCircularProgress-root-630 MuiCircularProgress-colorPrimary-633 MuiCircularProgress-indeterminate-632"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg-635"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle-636 MuiCircularProgress-circleIndeterminate-638"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
